### ISSUE TITLE
[Smallfix] Add custom pattern-matching ala Scala unapply

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#286]
I would like to be able to re-write this:

```
   case App(e1, e2) => switch {
        case isValue(e1) && isValue(e2) =>
            // An application where both arguments are values can be reduced.
            // Return the expression and the hole evaluation context.
            Some((e0, Hole))
```

into something where ` isValue(e1) && isValue(e2)` are used inside the pattern match. In other words, I would like to be able to define a function `isValue(e: Expression): Option[Value]` and use it like this:

```
   case App(isValue(e1), isValue(e2)) => ...
```

**Comment by mhyee** [comment]
How badly do we need/want this feature? I prefer to err on the side of keeping the language smaller and simpler, rather than adding too many features.

I never really had to define my own `unapply` in my Scala code. I think it was only once, in the early days of the Flix interpreter (and it has long since been refactored). How common is `unapply` in Scala? How common would this feature be in Flix?

**Comment by magnus-madsen** [comment]
I think user-defined unapply are rarely used in Scala, but I also think that the reason is that Scala pattern matching is a bit brittle, and their use can easily obscure the code. For example, you can't easily tell whether you are using an unapply or some case class, etc. 

If we allow you to write:

```
 case App(isValue(e1), isValue(e2)) => ...
```

then following our normal rules, `isValue` should be a function. It cannot be an enum, because enums must be uppercase.

On the other hand, how do we assign a meaning to these:

```
 case App(isValue(e1, e2)), isValue(e3)) => ...
 case App(isValue(Foo(x))), isValue(e2)) => ...
 case App(isValue(x + y)), isValue(e2)) => ...
```

I think 
(1) is fairly clear. Here `isValue` should return a pair which is bound as `(e1, e2)`.
(2) requires `isValue` to return an enum which has a constructor `Foo` which the result is matched against.
(3) is non-sense and should be disallowed.

In other words, in a pattern:

```
Foo(bar(x, y, z))
```

we should think of `bar` as a function from whatever `Foo` has a nested type (call it a) to an optional triple, i.e. `bar : a -> Option[(x, y, z)]`

**Comment by magnus-madsen** [comment]
For the record, the following should be valid (but may look weird to the untrained eye):

```
  case App(isValue()) => ..
  case App(isValue(_, _)) => ..
```

**Comment by magnus-madsen** [comment]
Since you asked about motivation, if grab a copy of e.g. this paper:

http://matt.might.net/papers/vanhorn2010abstract.pdf

and look at figure 1 you will see something very commonly used in semantics/abstract machines/program analysis. Namely, there is a bunch of transitions relations. The way they are written:

```
< v, \rho, ar(...) > ->
```

is actually doing a pattern match on a tuple where the first element is an ***expression*** but the rule is matching a ***value***. In essence, it is common to have two grammars, one is a subset of the other (e.g. expressions vs values) and you want to do pattern match on the sub grammar.

I will give you another example:

Would it not be nice if you could write:

```
enum Expr { ... }

def singleStep(e: Expr): Expr = match e with {
  case Plus(isValue(v1), isValue(v2)) => ...
  case Plus(isValue(v1), e) => ...
  case Plus(e1, _) => ...
}
```

and `v1` and `v2` could even have a different type than `Expr`!

**Comment by mhyee** [comment]
That's a good example, and the rules you proposed make sense. But I'm not convinced that Flix needs this feature, yet.

It doesn't allow us to express anything we can't already express, and it seems the main purpose is to make the code less verbose. But we can still do that if we implement #271:

```
enum Expr { ... }

def singleStep(e: Expr): Expr = match e with {
  case Plus(v1, v2) if isValue(v1) && isValue(v2) => ...
  case Plus(v1, e2) if isValue(v1) => ...
  case Plus(e1, _) => ...
}
```

I think this is a big improvement over nesting a switch inside a pattern match. It's slightly more verbose than custom pattern matching, but I think it's a lot simpler and clearer.

**Comment by magnus-madsen** [comment]
The only thing is that inside:

```
case Plus(v1, v2) if isValue(v1) && isValue(v2) => ...
```

You can't change the type of `v1`. So if you actually had a distinct enum `Value` you would have to write:

```
case Plus(v1, v2) if isValue(v1) && isValue(v2) => v1.toValue
```

 F# has something called active patterns which we should look at.

**Comment by magnus-madsen** [comment]
```f#
let (|Even|Odd|) input = if input % 2 = 0 then Even else Odd

let TestNumber input =
   match input with
   | Even -> printfn "%d is even" input
   | Odd -> printfn "%d is odd" input

TestNumber 7
TestNumber 11
TestNumber 32
```

```f#
//A very simple active pattern
let (|DiscountAuthorisationNeeded|DiscountApproved|) input = 
    if input > 10 then 
        DiscountAuthorisationNeeded else 
        DiscountApproved
    
let ApplyDiscount input =
    match input with
    | DiscountAuthorisationNeeded -> printfn "%d : Wil need to be authorised" input
    | DiscountApproved -> printfn "%d : Discount approved" input

ApplyDiscount 7
ApplyDiscount 11
ApplyDiscount 32
```

Looks interesting.

**Comment by magnus-madsen** [comment]
This mechanism could also be used to support pattern matching on sets and maps without extending the deep internals of the compiler.

**Comment by magnus-madsen** [comment]
https://ghc.haskell.org/trac/ghc/wiki/ViewPatterns
